src/func.cc

```diff
@@ -914,6 +916,150 @@ Value builtin_cross(const std::shared_ptr<Context> ctx, const std::shared_ptr<Ev
+    obj.set("true", 4);
+    obj.set("bool", true);
+    obj.set("str", "hello");
+    return std::move(obj);
+}
+
+Value builtin_textmetrics(const std::shared_ptr<Context> ctx, const std::shared_ptr<EvalContext> evalctx)
+{
+    AssignmentList args{
+        assignment("text"),
+        assignment("size"),
+        assignment("font")
+    };
+    AssignmentList optargs{
+        assignment("direction"),
+        assignment("language"),
+        assignment("script"),
+        assignment("halign"),
+        assignment("valign"),
+        assignment("spacing"),
+    };
+
+    ContextHandle<Context> c{Context::create<Context>(ctx)};
+    c->setVariables(evalctx, args, optargs);
+
+    // Note:
+    // This populates all of the Params entries that text() populates.
+    // Probably some of them (fn, fa, fs, segments) do not contribute
+    // to the text metrics.  However, we populate them here rather than
+    // "knowing" which ones do and do not contribute.
+    auto fn = c->lookup_variable("$fn").toDouble();
+    auto fa = c->lookup_variable("$fa").toDouble();
+    auto fs = c->lookup_variable("$fs").toDouble();
+
+    FreetypeRenderer::Params params;
+
+    params.set_loc(evalctx->loc);
+    params.set_documentPath(ctx->documentPath());
+    params.set_fn(fn);
+    params.set_fa(fa);
+    params.set_fs(fs);
+
+    auto size = c->lookup_variable_with_default("size", 10.0);
+    auto segments = Calc::get_fragments_from_r(size, fn, fs, fa);
+    // The curved segments of most fonts are relatively short, so
+    // by using a fraction of the number of full circle segments
+    // the resolution will be better matching the detail level of
+    // other objects.
+    auto text_segments = std::max(floor(segments / 8) + 1, 2.0);
+
+    params.set_size(size);
+    params.set_segments(text_segments);
+    params.set_text(c->lookup_variable_with_default("text", ""));
+    params.set_spacing(c->lookup_variable_with_default("spacing", 1.0));
+    params.set_font(c->lookup_variable_with_default("font", ""));
+    params.set_direction(c->lookup_variable_with_default("direction", ""));
+    params.set_language(c->lookup_variable_with_default("language", "en"));
+    params.set_script(c->lookup_variable_with_default("script", ""));
+    params.set_halign(c->lookup_variable_with_default("halign", "default"));
+    params.set_valign(c->lookup_variable_with_default("valign", "default"));
+
+    FreetypeRenderer renderer;
+    renderer.detect_properties(params);
+
+    FreetypeRenderer::TextMetrics metrics;
+    if (!metrics.calculate(params)) {
+        return Value::undefined.clone();
+    }
+
+    // The bounding box, ascent/descent, and offset values will be zero
+    // if the text consists of nothing but whitespace.
+    VectorType bbox_pos;
+    bbox_pos.emplace_back(metrics.bbox_x);
+    bbox_pos.emplace_back(metrics.bbox_y);
+
+    VectorType bbox_dims;
+    bbox_dims.emplace_back(metrics.bbox_w);
+    bbox_dims.emplace_back(metrics.bbox_h);
```

Comment:
Q:  should bounding boxes be represented as:
- [ left, right, bottom, top ]
- [ [minx, miny], [dimx, dimy] ]
- [ [minx, miny], [maxx, maxy] ]
?

Probably not the first one, because it's not a point or a dimension.  But:  it's equally inconvenient to get the coordinates of any of the four corners.

The second works well for things like a translate and a square.
The third works well for positioning to the lower left or upper right corner.

Translating between the second and third is just a matter of adding or subtracting.

Reply:
I vote the second, I can see resize() being used with this, having the width/height at hand is easier.
I presume that these are axis-aligned, and don't change with `direction=`?

Reply:
They change with `direction` in that the bounding boxes for horizontal text tend to be wide and not very tall, and for vertical text tend to be tall and not very wide.  But the position is always the lower-left corner, the minimum X/Y, and the dimension is always positive.  (Or, if the third variant, the second point always has X and Y greater or equal to the first point.)

It's a bounding box.  It's the minimum and maximum X and Y for the actual ink laid down.

Put another way, it's the translation and size of the smallest square() that will cover the ink laid down.